### DISCUSSION TITLE
Downloading language packs from play store not a good user experience

### DISCUSSION THREAD

**Discussion by archit2508** [AnySoftKeyboard/AnySoftKeyboard#4103] [Q&A] [answered]
@menny I am an Android developer and was trying to fork this repo and play with it to build my own version. It's a fabulous work you guys have done here. 

Is there a way the languages can be enabled directly from keyboard settings instead of downloading them from play store ? Downloading from play store is very troublesome. Is there any way we can, maybe store compiled APKs in the code itself and, maybe install them directly when user selects a lang. Or is there a way we can precompile all languages in the main app and just enable/disable them as per checkboxes ? Just like any other keyboard app does like Gboard, swiftkey, samsung keyboard etc.

I am keen to make this happen. Any help in this regard would be highly appreciated.

**Comment by menny** [comment, marked as answer]
That sounds great.

I will note that it is already possible to build the app with all languages prebundled in the app. If you look at the Gradle file of :ime:app you will see a variant called allAddOns. This, basically, adds all the packs as dependencies to the main app. 

Technically, the main app will scan for Android components that export addon capabilities. Whatever it finds it will expose to the user as an available pack. It doesn't have to be an external APK. 

Does this give you enough to start with?

**Reply by archit2508** [reply]
Yes @menny Thanks for the head start !

**Comment by ssvb** [comment]
@menny is it possible to do such allAddOns build entirely on Github CI and upload the resulting APKs as an artifact without installing Gradle on a local computer?

**Reply by menny** [reply]
That will be a very large APK.
what is the use case?

**Reply by ssvb** [reply]
How large would it be? The use case might be a much easier testing for those, who are not Android developers, such as https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/4225

**Reply by menny** [reply]
I tested that with the latest code and it is 124061746 bytes, or 124MB. That's not a small thing to build and keep (we do several builds a week)

**Reply by krystalcode** [reply]
Some users also do not use a store like Play Store or F-Droid, rather download APKs directly from the developers. It would be good to have a method for downloading the languages we want directly from here.
